src/stores/blockstore.js

```diff
@@ -0,0 +1,139 @@
+'use strict'
+
+const Block = require('ipfs-block')
+const pull = require('pull-stream')
+const Lock = require('lock')
+const base32 = require('base32.js')
+const path = require('path')
+const write = require('pull-write')
+const parallel = require('run-parallel')
+const through = require('pull-through')
+
+const PREFIX_LENGTH = 5
+
+exports = module.exports
+
+function multihashToPath (multihash, extension) {
+  extension = extension || 'data'
+  const encoder = new base32.Encoder()
+  const hash = encoder.write(multihash).finalize()
+  const filename = `${hash}.${extension}`
+  const folder = filename.slice(0, PREFIX_LENGTH)
+
+  return path.join(folder, filename)
+}
+
+exports.setUp = (basePath, BlobStore, locks) => {
+  const store = new BlobStore(basePath + '/blocks')
+  const lock = new Lock()
+
+  function writeBlock (block, cb) {
+    if (!block || !block.data) {
+      return cb(new Error('Invalid block'))
+    }
+
+    const key = multihashToPath(block.key, block.extension)
+
+    lock(key, (release) => pull(
+      pull.values([block.data]),
+      store.write(key, release((err) => {
+        if (err) {
+          return cb(err)
+        }
+        cb(null, {key})
+      }))
+    ))
+  }
+
+  return {
+    getStream (key, extension) {
+      extension = extension || 'data'
+
+      if (!key) {
+        return pull.error(new Error('Invalid key'))
+      }
+
+      const p = multihashToPath(key, extension)
+
+      const ext = extension === 'data' ? 'protobuf' : extension
+      let data = []
+
+      return pull(
+        store.read(p),
+        through(function (values) {
+          data = data.concat(values)
+        }, function () {
+          this.queue(new Block(Buffer.concat(data), ext))
+          this.queue(null)
+        })
+      )
+    },
+
+    putStream () {
+      let ended = false
+      let written = []
+      let push = null
+
+      const sink = write((blocks, cb) => {
+        parallel(blocks.map((block) => (cb) => {
+          writeBlock(block, (err, meta) => {
+            if (err) return cb(err)
+            if (push) {
+              const read = push
+              push = null
+              read(null, meta)
+              return cb()
+            }
+
+            written.push(meta)
+            cb()
+          })
+        }), cb)
+      }, null, 100, (err) => {
+        ended = err || true
+        if (push) push(ended)
+      })
+
+      const source = (end, cb) => {
+        if (end) ended = end
+        if (ended) return cb(ended)
+
+        if (written.length) {
+          return cb(null, written.shift())
+        }
+
+        push = cb
+      }
+
+      return {source, sink}
+    },
+
+    has (key, extension, cb) {
+      if (typeof extension === 'function') {
+        cb = extension
+        extension = undefined
```

Comment:
In `get` we default to `data`, but in the remaining ops we don't.

```
const ext = extension === 'data' ? 'protobuf' : extension
```

Why is that? Any rationale?